airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/DetectStreamToFlush.java

```diff
@@ -0,0 +1,156 @@
+   * zero. This helps in the case where there are a lot of streams, so total memory usage is high, but
+   * each individual queue isn't that large.</li>
+   * <li>closing - If the Flush Worker is closing, we reduce it to zero. We close when all records
+   * have been added to the queue, at which point, our goal is to flush out any non-empty queues.</li>
+   *
+   * @return based on the conditions, the threshold in bytes.
+   */
+  @VisibleForTesting
+  long computeQueueThreshold() {
+    final boolean isBuffer90Full =
+        EAGER_FLUSH_THRESHOLD <= (double) bufferDequeue.getTotalGlobalQueueSizeBytes() / bufferDequeue.getMaxQueueSizeBytes();
+    // when we are closing or queues are very fully, flush regardless of how few items are in the queue.
+    return isClosing.get() || isBuffer90Full ? 0 : QUEUE_FLUSH_THRESHOLD_BYTES;
+  }
+
+  // todo (cgardens) - improve prioritization by getting a better estimate of how much data running
+  // workers will process. we have access to their batch sizes after all!
+  /**
+   * Iterates over streams until it finds one that is ready to flush. Streams are ordered by priority.
+   * If no stream is ready, it returns an empty optional.
+   * <p>
+   * A stream is ready to flush if it either meets a size threshold or a time threshold. For the size
+   * threshold, the size of the data in the queue is compared to the threshold that is passed into
+   * this method.
+   * <p>
+   * One caveat, is that if that stream already has a worker running, we "penalize" its size. We do
+   * this by computing what the size of the queue would be after the running workers for that queue
+   * complete. This is based on a dumb estimate of how much data a worker can process. There is an
+   * opportunity for optimization here, by being smarter about predicting how much data a running
+   * worker is likely to process.
+   * <p>
+   * Finally, the time trigger is based on the last time a record was added to the queue. We don't
+   * want records to sit forever, even if the queue is not that full (bad for time to value for
+   * users). Also, the more time passes since a record was added, the less likely another record is
+   * coming (caveat is CDC where it's random).
+   *
+   * @param queueSizeThresholdBytes - the size threshold to use for determining if a stream is ready
+   *        to flush.
+   * @return the next stream to flush. if no stream is ready to flush, empty.
+   */
+  @VisibleForTesting
+  Optional<StreamDescriptor> getNextStreamToFlush(final long queueSizeThresholdBytes) {
+    for (final StreamDescriptor stream : orderStreamsByPriority(bufferDequeue.getBufferedStreams())) {
+      // while we allow out-of-order processing for speed improvements via multiple workers reading from
+      // the same queue, also avoid scheduling more workers than what is already in progress.
+      final long runningBytesEstimate = runningFlushWorkers.getNumFlushWorkers(stream) * QUEUE_FLUSH_THRESHOLD_BYTES;
```

Comment:
non-blocking thought - if the optimal batch size is many times larger than the QUEUE_FLUSH_THRESHOLD_BYTES, we can end up queueing up flush workers with small batches.

possible solution: a flush worker aborts if the queue it's pulling from is less than the flush size?

Reply:
agreed. that's the thing i want to layer on in the next pr. wanted to get some tests down first though.